homeassistant/components/bme280/sensor.py

```diff
@@ -1,153 +1,134 @@
+        return
     SENSOR_TYPES[SENSOR_TEMP][1] = hass.config.units.temperature_unit
-    name = config[CONF_NAME]
-    i2c_address = config[CONF_I2C_ADDRESS]
-
-    bus = smbus.SMBus(config[CONF_I2C_BUS])
-    sensor = await hass.async_add_executor_job(
-        partial(
-            BME280,
-            bus,
-            i2c_address,
-            osrs_t=config[CONF_OVERSAMPLING_TEMP],
-            osrs_p=config[CONF_OVERSAMPLING_PRES],
-            osrs_h=config[CONF_OVERSAMPLING_HUM],
-            mode=config[CONF_OPERATION_MODE],
-            t_sb=config[CONF_T_STANDBY],
-            filter_mode=config[CONF_FILTER_MODE],
-            delta_temp=config[CONF_DELTA_TEMP],
-            logger=_LOGGER,
+    sensor_conf = discovery_info[SENSOR_DOMAIN]
+    name = sensor_conf[CONF_NAME]
+    scan_interval = sensor_conf[CONF_SCAN_INTERVAL]
+    if CONF_SPI_BUS in sensor_conf and CONF_SPI_DEV in sensor_conf:
+        spi_dev = sensor_conf[CONF_SPI_DEV]
+        spi_bus = sensor_conf[CONF_SPI_BUS]
+        _LOGGER.debug("BME280 sensor initialize at %s.%s", spi_bus, spi_dev)
+        sensor = await hass.async_add_executor_job(
+            partial(
+                BME280_spi,
+                t_mode=sensor_conf[CONF_OVERSAMPLING_TEMP],
+                p_mode=sensor_conf[CONF_OVERSAMPLING_PRES],
+                h_mode=sensor_conf[CONF_OVERSAMPLING_HUM],
+                standby=sensor_conf[CONF_T_STANDBY],
+                filter=sensor_conf[CONF_FILTER_MODE],
+                spi_bus=sensor_conf[CONF_SPI_BUS],
+                spi_dev=sensor_conf[CONF_SPI_DEV],
+            )
         )
+        if not sensor.sample_ok:
+            _LOGGER.error("BME280 sensor not detected at %s.%s", spi_bus, spi_dev)
+            return
+    else:
+        i2c_address = sensor_conf[CONF_I2C_ADDRESS]
+        bus = smbus.SMBus(sensor_conf[CONF_I2C_BUS])
+        sensor = await hass.async_add_executor_job(
+            partial(
+                BME280_i2c,
+                bus,
+                i2c_address,
+                osrs_t=sensor_conf[CONF_OVERSAMPLING_TEMP],
+                osrs_p=sensor_conf[CONF_OVERSAMPLING_PRES],
+                osrs_h=sensor_conf[CONF_OVERSAMPLING_HUM],
+                mode=sensor_conf[CONF_OPERATION_MODE],
+                t_sb=sensor_conf[CONF_T_STANDBY],
+                filter_mode=sensor_conf[CONF_FILTER_MODE],
+                delta_temp=sensor_conf[CONF_DELTA_TEMP],
+            )
+        )
+        if not sensor.sample_ok:
+            _LOGGER.error("BME280 sensor not detected at %s", i2c_address)
+            return
+
+    async def async_update_data():
+        await hass.async_add_executor_job(sensor.update)
+        if not sensor.sample_ok:
+            raise UpdateFailed(f"Bad update of sensor {name}")
+        return sensor
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name=DOMAIN,
+        update_method=async_update_data,
+        update_interval=scan_interval,
```

Comment:
We should cap the minimum scan interval to the existing minimum time between updates.